Force app update

leofelix077 opened the pull request:
Closes #451 

Amplitude doesn't allow for dots, so I set as underscore. Open for suggestions

We could also change the "I'll update later" to "Exit application" and force users to update, just wasn't sure about the consequences coming with it (in the sense of locking user out of the app if they can't update. meaning they could potentially not be able to access their funds for a while. but behavior can also be changed)

Also added analytics events for opening the app stores and dismissing the full screen notice

### Moved this video to the top to show the confirmation modal. @CassioMG  and @sdfcharles -> Not sure if we want to to double confirmation, but here's the shell for it. Copy TBD. if not, let me know, then I can remove it. The other videos do not have the confirmation modal

https://github.com/user-attachments/assets/bf275e1e-8dcf-44b6-8b6e-ce26ec03f261


<img width="300" alt="Screenshot 2025-10-16 at 15 10 13" src="https://github.com/user-attachments/assets/01fb168b-40bb-4164-a6e6-07d146eab01e" />
<img width="300" alt="Screenshot 2025-10-16 at 15 10 28" src="https://github.com/user-attachments/assets/e6665f76-e79b-440a-a173-f9b308382372" />

Setting the payload strings on the update texts

<img width="700"  alt="Screenshot 2025-10-16 at 15 10 21" src="https://github.com/user-attachments/assets/97adbcef-3b7f-481a-9ed0-7a6db39c8f07" />


## All iOS versions do not redirect to the app store on the Simulator (not available). Further checks must be done on real device afterwards


iOS Setting Optional and Mandatory flags with analytics enabled (to test the no-interfence with the normal flag)

https://github.com/user-attachments/assets/812e7c81-99e0-4258-9ebc-9d4950e19a5b

iOS Switching override to current version on debug screen -> to remove the banner

https://github.com/user-attachments/assets/03db2256-586c-419c-b98a-bef3bf3524b1

iOS Switching from current version to old app version to trigger full screen:

https://github.com/user-attachments/assets/f4183194-cb93-42e2-814a-c6c93b6e8be9

iOS Testing forcing an error to show the toast (in case Linking fails. shouldn't happen)

https://github.com/user-attachments/assets/ede3939f-54b1-4a55-8e26-74b4c65a4474

iOS Unauthenticated user show full screen on app open

https://github.com/user-attachments/assets/bbd8bacc-eec2-4183-8b71-a7ba94930ecd

iOS Analytics disabled on settings -> trigger full screen app update and modal

https://github.com/user-attachments/assets/5c01ca00-15cb-4230-80d4-46811045f685

iOS Show App update screen even if on start screen (e.g. old app version installation and never used)

https://github.com/user-attachments/assets/f3b30f29-e2ff-4558-82b4-73f38a8177af


Android basic flow of showing banner and full screen depending on app version

https://github.com/user-attachments/assets/65c7cb5e-09d8-45e7-9c18-e7d2e1026de0

Android: Full Screen and Banner showing on app reload, and when logged out

https://github.com/user-attachments/assets/1dae02c1-4ff4-427f-aece-2034b570dd17

Android: Logged out / Reload / Welcome Screen banner and full screen without Analytics enabled

https://github.com/user-attachments/assets/70bea3eb-be76-4958-aa6a-6d692a336889


### What

Add ability to force app updates from remote, by setting the latest app version and setting the minimum required version

### Why

When a critical update is rolled out, we want to be able to force users to update the app

### Known limitations

## 1. iOS doesn't have an easy way to redirect to TestFlight, so it needs to be tested on production builds 

## 2. In-app updates, as seen below, will be added later - https://github.com/stellar/freighter-mobile/issues/488

![In app update example](https://user-images.githubusercontent.com/8539174/88419625-6db0ef00-cddd-11ea-814e-389db852368b.gif)

### Checklist

#### PR structure

- [x] This PR does not mix refactoring changes with feature changes (break it down into smaller PRs if not).
- [x] This PR has reasonably narrow scope (break it down into smaller PRs if not).
- [x] This PR includes relevant before and after screenshots/videos highlighting these changes.
- [x] I took the time to review my own PR.

#### Testing

- [x] These changes have been tested and confirmed to work as intended on Android.
- [x] These changes have been tested and confirmed to work as intended on iOS.
- [x] These changes have been tested and confirmed to work as intended on small iOS screens.
- [x] These changes have been tested and confirmed to work as intended on small Android screens.
- [x] I have tried to break these changes while extensively testing them.
- [x] This PR adds tests for the new functionality or fixes.

#### Release

- [x] This is not a breaking change.
- [x] This PR updates existing JSDocs when applicable.
- [x] This PR adds JSDocs to new functionalities.
- [ ] I've checked with the product team if we should add metrics to these changes.
- [x] I've shared relevant before and after screenshots/videos highlighting these changes with the design team and they've approved the changes.

CassioMG commented on src/hooks/useAppUpdate.ts:
```diff
@@ -0,0 +1,97 @@
+import { Linking } from "react-native";
+import { getBundleId, getVersion } from "react-native-device-info";
+
+const IOS_BUNDLE_ID = "id6743947720";
```
```suggestion
const IOS_APPLE_ID = "id6743947720";
```

just to reduce confusion as Bundle Id actually is `org.stellar.freighterwallet` but the Apple ID is `6743947720`

CassioMG commented on src/ducks/remoteConfig.ts:
```diff
@@ -1,26 +1,62 @@
+const STRING_FLAGS = ["required_app_version", "latest_app_version"] as const;
+
+// Complex feature flags configuration (objects, JSON, etc.)
+const COMPLEX_FLAGS = ["app_update_text"] as const;
```
would there be a way to automate those array creations based on the `BooleanFeatureFlags`, `StringFeatureFlags` and `ComplexFeatureFlags` keys? So devs don't need to remember to add new flags to both places

CassioMG commented on src/components/screens/ForceUpdateScreen/ForceUpdateScreen.tsx:
```diff
@@ -0,0 +1,83 @@
+
+/**
+ * Full-screen component that forces users to update their app
+ * Used for critical updates that could affect fund security
```
```suggestion
 * Used for critical updates or when users are too far behind on app version
```

CassioMG commented on src/hooks/useAppUpdate.ts:
```diff
@@ -0,0 +1,97 @@
+    isDev && overriddenAppVersion ? overriddenAppVersion : getVersion();
+
+  // Parse the update text JSON for internationalization
+  const getUpdateMessage = useCallback(() => {
```
should we use the external helper here?

CassioMG commented on src/helpers/versionComparison.ts:
```diff
@@ -0,0 +1,80 @@
+ * @param targetVersion - Target version to compare against
+ * @returns Object with major, minor, and patch differences
+ */
+export const getVersionDifference = (
```
this is based on what we have in the [issue description](https://github.com/stellar/freighter-mobile/issues/451)

CassioMG replied:
oh yeah, you are right, let's use the flags instead of hardcoding it.

I think we can remove this function then? It's not being used anywhere

leofelix077 commented:
iOS Test minor/major/protocol changes (Required app version set on amplitude to 1.4.23)

https://github.com/user-attachments/assets/a0b77232-ce61-4300-b37b-6f8ba21b5659

CassioMG commented on src/hooks/useAppUpdate.ts:
```diff
@@ -35,31 +32,16 @@ export const useAppUpdate = () => {
+  const updateMessage = getAppUpdateText(updateText);
 
   // Only check for updates when remote config is initialized
+  // For this app, any version difference (including protocol) requires a forced update
```
let's remove this comment (and changes) and use the flags to decide which app update should be "required" or "optional" like the PR was originally doing - thanks!

leofelix077 commented:
@CassioMG oh yes true. I just wanted to add it to make explicit, as I assume protocol would be breaking changes, but is still covered by the requiredAppVersion

reverted now and added a unit test for this scenario

CassioMG commented:
> @CassioMG oh yes true. I just wanted to add it to make explicit, as I assume protocol would be breaking changes, but is still covered by the requiredAppVersion
> 
> reverted now and added a unit test for this scenario

Thanks. Protocol changes are not always breaking changes, but it's strongly advised that everyone upgrade to it whenever there is a new version

leofelix077 commented:
Test for version specific flag setting so it can be shown again to the same user if they fall behind on versions again with future updates

https://github.com/user-attachments/assets/916ed7ce-e735-4ee5-a17d-e2b997a8e7cf



iOS Normal flow / not showing banner during same session
https://github.com/user-attachments/assets/b7b624e2-d20f-47da-ad7d-041ef01ad328

CassioMG commented on src/helpers/appUpdateText.ts:
```diff
@@ -0,0 +1,27 @@
+ * @param appUpdateText - The app update text object with enabled and payload
+ * @returns The localized update text
+ */
+export const getAppUpdateText = (appUpdateText: {
```
would you mind adding a video showing the text being updated through the feature flags? Maybe switching between short and long texts - thanks!

leofelix077 commented on src/config/envConfig.ts:
```diff
@@ -89,7 +89,7 @@ const getWalletKitConfig = (isProduction: boolean) => {
  */
 const getEnvConfig = (): EnvConfigType => ({
   // Let's avoid sending events to Amplitude while developing locally
-  AMPLITUDE_API_KEY: __DEV__ ? "" : Config.AMPLITUDE_API_KEY,
```
oh no. this was for testing. thanks!

CassioMG reviewed (approved):
LGTM
